### PR TITLE
[pytorch] include all overloads for OSS custom build

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42214 [pytorch] include all overloads for OSS custom build**

Summary:
For mobile custom build, we only generate code for ops that are used by
specific models to reduce binary size.

There multiple places where we apply the op filtering:
- generated_unboxing_wrappers_*.cpp
- autograd/VariableType*.cpp
- c10 op registration (in aten/gen.py)

For c10 op registration, we filter by the main op name - all overloads
that match the main op name part will be kept.

For generated_unboxing_wrappers_*, we filter by the full op name - only
those having exactly the same overload name will be kept.

This PR changes generated_unboxing_wrappers_* and autograd/VariableType*.cpp
codegen to also filter by the main op name.

The reasons are:
- keeping all overloads can have better backward compatibility;
- generated_unboxing_wrappers_* are relatively small as it only contains
  thin wrappers for root ops.
- generated_unboxing_wrappers_* will be replaced by c10 op registration
  soon anyway.
- autograd/VariableType*.cpp are not included in OSS build.

Why it offers better backward compatibility? #40737 is an example:
It introduced a new `_convolution` overload and renamed the original one
to `_convolution.deprecated`. Before this PR, the model prepared by the
old version PyTorch won't be able to run on the custom mobile build
generated on the PR because `_convolution.deprecated` won't be kept in
the custom build due to full op name matching policy. By relaxing it to
partial matching policy, the mobile custom build CI on the PR can pass.

Will test the size impact for FB production build before landing.

Differential Revision: [D22809564](https://our.internmc.facebook.com/intern/diff/D22809564)

### CODE REVIEW

File: tools/jit/gen_unboxing_wrappers.py [hunk 1]
```diff
@@ -398,8 +392,8 @@ def filter_decls(jit_decls, disable_autograd, selected_op_list, force_schema_reg
         for decl in jit_decls:
             if disable_autograd and is_backward_op(decl):
                 continue
-            op_name = signature_without_args(decl)
-            if selected_op_list and op_name not in selected_op_list:
+            op_name = op_name_without_overload(decl)
+            if selected_op_list is not None and op_name not in selected_op_list:
```
Comment (iseeyuan) [inline tools/jit/gen_unboxing_wrappers.py:396]: Good catch!

**Review by iseeyuan — APPROVED** [review]
LGTM!

**Comment by ezyang** [comment]
One question for my information: wouldn't #40737 still have broken BC because it changed the overload name of an operator? I was under the impression that overload name mattered too?

**Comment by iseeyuan** [comment]
> One question for my information: wouldn't #40737 still have broken BC because it changed the overload name of an operator? I was under the impression that overload name mattered too?

I believe this fix is for OSS custom build, which is using full-jit to handle BC. Yes #40737 still breaks BC for mobile. I'm working on a fix from bytecode side and will share it with you when the PR is ready.

**Comment by ljk53** [comment]
> One question for my information: wouldn't #40737 still have broken BC because it changed the overload name of an operator? I was under the impression that overload name mattered too?

Right, #40737 breaks two things: 1) OSS mobile selective build; 2) FB lite-interpreter buid;
This PR fixes 1); Martin is fixing 2) by adding an "adapter" to lite-interpreter to support old schema;

@ezyang I haven't checked why we renamed the original schema's overload to "deprecated" instead of simply deleting it. Cannot new schema + default value handle BC problem for full-JIT?
And if we do want to keep the original _convolution maybe we can keep the empty overload and give a new overload (e.g. "tf32" or "v2"?) to the new schema? This way maybe we can workaround 2) for now...

**Comment by ljk53** [comment]
closed in https://github.com/pytorch/pytorch/commit/a01e91e6b2ab28b000d66588b9d32a91f8fc4e62

**Comment by zasdfgbnm** [comment]
new schema + default value does not sound like a good solution because the default value can neither be true nor false, but it should depend on a global flag. Using a `bool?=None` might be a workaround, but this makes the signature confusing on why `allow_tf32` are handled differently from all the other flags.

And isn't adding a `.deprecated` to a deprecated operator the recommended way to handle deprecation? https://github.com/pytorch/pytorch/pull/31524#discussion_r360586263

**Comment by ezyang** [comment]
> Using a bool?=None might be a workaround, but this makes the signature confusing on why allow_tf32 are handled differently from all the other flags.

As the person responsible for the `_convolution` signature back in the day, I can say: I wrote this signature in a simpler time when there wasn't anything such as JIT BC concerns and I just wanted a purely functional version of the signature where you could explicitly specify all inputs without relying on any global state. That's it. In the world of JIT BC it's probably wrong to have signatures like this actually show up in the serialized models, since it makes it harder to maintain BC.

**Comment by ngimel** [comment]
> As the person responsible for the `_convolution` signature back in the day, I can say: I wrote this signature in a simpler time when there wasn't anything such as JIT BC concerns and I just wanted a purely functional version of the signature 

TIL It's no longer functional, global state for mkldnn enablement is queried inside `_convolution`

**Pull request closed without merging**